osis-core/src/main/java/com/scality/osis/service/impl/ScalityOsisServiceImpl.java

```diff
@@ -1042,7 +1040,46 @@ public ScalityOsisCaps updateOsisCaps(ScalityOsisCaps osisCaps) {
 
     @Override
     public PageOfOsisBucketMeta getBucketList(String tenantId, long offset, long limit) {
-        throw new NotImplementedException();
+        try {
+            logger.info("Get bucket list request received:: tenant ID:{}, offset:{}, limit:{}", tenantId, offset, limit);
+            Credentials tempCredentials = getCredentials(tenantId);
+            final AmazonS3 s3Client = this.s3.getS3Client(tempCredentials,
+                    appEnv.getRegionInfo().get(0));
+
+            //s3 listBucket has no pagination, so list all
+            List<Bucket> buckets = s3Client.listBuckets();
+
+            logger.debug("[S3] List all Buckets response:{}", new Gson().toJson(buckets));
+
+            PageOfOsisBucketMeta pageOfOsisBucketMeta = ScalityModelConverter.toPageOfOsisBucketMeta(buckets, offset, limit);
+
+            logger.info("List Buckets response:{}", new Gson().toJson(pageOfOsisBucketMeta));
+
+            return pageOfOsisBucketMeta;
+        } catch (Exception e) {
+
+            if (isAdminPolicyError(e)) {
+                try {
+                    generateAdminPolicy(tenantId);
+                    return getBucketList(tenantId, offset, limit);
```

Comment:
Is it possible that this line is called everytime, in case we have consistent adminPolicy errors? This looks recursive, just want to be sure that we'll never end up in an infinite loop.

Reply:
indeed...thinking if there is a better way to do this